subsys/net/lib/dns/dns_sd.c

```diff
@@ -0,0 +1,786 @@
+		   uint16_t *host_offset)
+{
+	uint16_t total_size;
+	struct dns_rr *rr;
+	struct dns_srv_rdata *rdata;
+	size_t label_size;
+	uint16_t inst_offs;
+	uint16_t offset = buf_offset;
+
+	if ((DNS_SD_PTR_MASK & instance_offset) != 0) {
+		LOG_ERR("offset %u too big for message compression",
+			instance_offset);
+		return -E2BIG;
+	}
+
+	if ((DNS_SD_PTR_MASK & domain_offset) != 0) {
+		LOG_ERR("offset %u too big for message compression",
+			domain_offset);
+		return -E2BIG;
+	}
+
+	/* First, calculate that there is enough space in the buffer */
+	total_size =
+		/* pointer to .<Instance>.<Service>.<Protocol>.local. */
+		DNS_POINTER_SIZE + sizeof(*rr)
+		+ sizeof(*rdata)
+		/* .<Instance> */
+		+ DNS_LABEL_LEN_SIZE
+		+ strlen(inst->instance)
+		/* pointer to .local. */
+		+ DNS_POINTER_SIZE;
+
+	if (offset > buf_size || total_size >= buf_size - offset) {
+		LOG_ERR("Buffer too small. required: %u available: %d",
+			total_size, (int)buf_size - (int)offset);
+		return -ENOSPC;
+	}
+
+	/* insert a pointer to the instance + service name */
+	inst_offs = instance_offset;
+	inst_offs |= DNS_SD_PTR_MASK;
+	inst_offs = htons(inst_offs);
+	memcpy(&buf[offset], &inst_offs, sizeof(inst_offs));
+	offset += sizeof(inst_offs);
+
+	rr = (struct dns_rr *)&buf[offset];
+	rr->type = htons(DNS_RR_TYPE_SRV);
+	rr->class_ = htons(DNS_CLASS_IN | DNS_CLASS_FLUSH);
+	rr->ttl = htonl(ttl);
+	/* .<Instance>.local. */
+	rr->rdlength = htons(sizeof(*rdata) + DNS_LABEL_LEN_SIZE
+			     + strlen(inst->instance) +
+			     DNS_POINTER_SIZE);
+	offset += sizeof(*rr);
+
+	rdata = (struct dns_srv_rdata *)&buf[offset];
+	rdata->priority = 0;
+	rdata->weight = 0;
+	rdata->port = htons(inst->port);
+	offset += sizeof(*rdata);
+
+	*host_offset = offset;
+
+	label_size = strlen(inst->instance);
+	buf[offset++] = label_size;
+	memcpy(&buf[offset], inst->instance, label_size);
+	offset += label_size;
+
+	domain_offset |= DNS_SD_PTR_MASK;
+	domain_offset = htons(domain_offset);
+	memcpy(&buf[offset], &domain_offset, sizeof(domain_offset));
+	offset += sizeof(domain_offset);
+
+	__ASSERT_NO_MSG(total_size == offset - buf_offset);
+
+	return offset - buf_offset;
+}
+
+int dns_sd_handle_ptr_query(const struct dns_sd_rec *inst, bool ipv6,
+			    const uint8_t *addr,
+			    uint8_t *buf, uint16_t buf_size)
+{
+	/*
+	 * RFC 6763 Section 12.1
+	 *
+	 * When including a DNS-SD Service Instance Enumeration or Selective
+	 * Instance Enumeration (subtype) PTR record in a response packet, the
+	 * server/responder SHOULD include the following additional records:
+	 *
+	 * o  The SRV record(s) named in the PTR rdata.
+	 * o  The TXT record(s) named in the PTR rdata.
+	 * o  All address records (type "A" and "AAAA") named in the SRV rdata.
+	 *	contain the SRV record(s), the TXT record(s), and the address
+	 *      records (A or AAAA)
+	 */
+
+	uint16_t instance_offset = -1;
+	uint16_t service_offset = -1;
+	uint16_t domain_offset = -1;
+	uint16_t host_offset = -1;
```

Comment:
Setting the unsigned variables to -1 looks suspicious, could you elaborate why to do it like this?

Reply:
Possibly a copy-paste from my test harness. They don't need to be initialized.